File: javascript/packages/highlighter/src/document.ts
import type { DiagnosticSeverity } from "@herb-tools/core"

export type StyleRole =
  | { kind: "Plain" }
  | { kind: "Token"; tokenType: string }
  | { kind: "RubyKeyword" }
  | { kind: "TagName" }
  | { kind: "AttributeName" }
  | { kind: "AttributeValue" }
  | { kind: "CommentInterior" }

export interface StyledRun {
  text: string
  role: StyleRole
}

export interface Document {
  version: 1
  nodes: Node[]
}

export interface FileHeaderNode {
  type: "FileHeader"
  path: string
  line: number | null
  column: number | null
  url: string | null
}

export interface DiagnosticHeaderNode {
  type: "DiagnosticHeader"
  severity: DiagnosticSeverity
  message: string
  code: string | null
  codeUrl: string | null
  suffix: string | null
}

export interface CodeBlockNode {
  type: "CodeBlock"
  kind: CodeBlockKind
  firstLine: number
  runs: StyledRun[]
  lines: LineInfo[]
}

export interface ProgressRuleNode {
  type: "ProgressRule"
  index: number
  total: number
}

export type Node =
  | FileHeaderNode
  | DiagnosticHeaderNode
  | CodeBlockNode
  | ProgressRuleNode

export type CodeBlockKind = "Listing" | "AnnotatedListing" | "Excerpt"

export interface LineInfo {
  number: number
  emphasis: LineEmphasis
  annotations: Annotation[]
}

export type LineEmphasis =
  | { kind: "Normal" }
  | { kind: "Dimmed" }
  | { kind: "Focus" }
  | { kind: "Marked"; severity: DiagnosticSeverity }

export interface Annotation {
  start: number
  end: number
  severity: DiagnosticSeverity
  message: AnnotationMessage | null
}

export interface AnnotationMessage {
  text: string
  code: string | null
  codeUrl: string | null
}

File: rust/herb-highlighter/src/document.rs
use serde::ser::SerializeMap;
use serde::{Serialize, Serializer};

use crate::diagnostic::DiagnosticSeverity;

#[derive(Debug, Clone, PartialEq)]
pub enum StyleRole {
  Plain,
  Token(String),
  RubyKeyword,
  TagName,
  AttributeName,
  AttributeValue,
  CommentInterior,
}

impl StyleRole {
  pub fn kind(&self) -> &'static str {
    match self {
      StyleRole::Plain => "Plain",
      StyleRole::Token(_) => "Token",
      StyleRole::RubyKeyword => "RubyKeyword",
      StyleRole::TagName => "TagName",
      StyleRole::AttributeName => "AttributeName",
      StyleRole::AttributeValue => "AttributeValue",
      StyleRole::CommentInterior => "CommentInterior",
    }
  }
}

impl Serialize for StyleRole {
  fn serialize<S: Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
    let mut map = serializer.serialize_map(None)?;

    map.serialize_entry("kind", self.kind())?;

    if let StyleRole::Token(token_type) = self {
      map.serialize_entry("tokenType", token_type)?;
    }

    map.end()
  }
}

#[derive(Debug, Clone, PartialEq, Serialize)]
pub struct StyledRun {
  pub text: String,
  pub role: StyleRole,
}

#[derive(Debug, Clone, PartialEq, Serialize)]
pub struct Document {
  pub version: u32,
  pub nodes: Vec<Node>,
}

#[derive(Debug, Clone, PartialEq, Serialize)]
#[serde(tag = "type")]
pub enum Node {
  FileHeader {
    path: String,
    line: Option<usize>,
    column: Option<usize>,
    url: Option<String>,
  },

  DiagnosticHeader {
    severity: DiagnosticSeverity,
    message: String,
    code: Option<String>,
    #[serde(rename = "codeUrl")]
    code_url: Option<String>,
    suffix: Option<String>,
  },

  CodeBlock {
    kind: CodeBlockKind,
    #[serde(rename = "firstLine")]
    first_line: usize,
    runs: Vec<StyledRun>,
    lines: Vec<LineInfo>,
  },

  ProgressRule {
    index: usize,
    total: usize,
  },
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize)]
pub enum CodeBlockKind {
  Listing,
  AnnotatedListing,
  Excerpt,
}

#[derive(Debug, Clone, PartialEq, Serialize)]
pub struct LineInfo {
  pub number: usize,
  pub emphasis: LineEmphasis,
  pub annotations: Vec<Annotation>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize)]
#[serde(tag = "kind")]
pub enum LineEmphasis {
  Normal,
  Dimmed,
  Focus,
  Marked { severity: DiagnosticSeverity },
}

#[derive(Debug, Clone, PartialEq, Serialize)]
pub struct Annotation {
  pub start: usize,
  pub end: usize,
  pub severity: DiagnosticSeverity,
  pub message: Option<AnnotationMessage>,
}

#[derive(Debug, Clone, PartialEq, Serialize)]
pub struct AnnotationMessage {
  pub text: String,
  pub code: Option<String>,
  #[serde(rename = "codeUrl")]
  pub code_url: Option<String>,
}
